### PR TITLE
Add inference and more

### PR DESCRIPTION
**Pull request by mayaEidelman**

### CODE REVIEW

File: config/TimeMixerPP_SCAT.yaml
Comment (SharonPeled) [inline config/TimeMixerPP_SCAT.yaml:1]: FYI we don't really use TimeMixer anymore, I think I'll remove it altogether soon

File: config/PatchTransformer_SCAT.yaml
Comment (SharonPeled) [inline config/PatchTransformer_SCAT.yaml:1]: Isn't the detection metrics part of the test?

File: main.py
Comment (SharonPeled) [inline main.py:1]: why we need both run_inference and run_detection? Isn't this the same thing?

File: src/data/AbstractDataset.py [hunk 1]
```diff
@@ -26,14 +26,23 @@ def __getitem__(self, idx):
...
+        # df_trajectory.drop(['y_detected'], axis=1, errors='ignore', inplace=True)
+
+        # y_detected = torch.tensor(df_trajectory['y_detected'].values).float()
+        y_detected = torch.randint(0, 2, (len(df_trajectory),)).float()
         
         ts = torch.tensor(df_trajectory.values).float()
         nan_mask = torch.isnan(ts).any(-1)  # Create a mask where values are NaN
 
-        sample = {'ts': ts, 'nan_mask': nan_mask, 'path': file_path, 'columns': list(df_trajectory.columns), 'timestamps': timestamps}
+        sample = {'ts': ts, 'nan_mask': nan_mask, 'path': file_path, 'columns': list(df_trajectory.columns), 'timestamps': timestamps, 'y_detected': y_detected}
```
Comment (SharonPeled) [inline src/data/AbstractDataset.py:45]: y_anomaly maybe?

File: src/data/GeneralTrajectoryDataModule.py
Comment (SharonPeled) [inline src/data/GeneralTrajectoryDataModule.py:1]: I don't think you need to implement predict_dataloader, you can use test loader instead 
In inference (where you use predict), you generally dont have the y, so you can't really log things

File: src/infer.py
Comment (SharonPeled) [inline src/infer.py:1]: I don't think we need to implement a new process for anomaly detection metrics
Ideally, it should be part of the test process where we have anomaly labels. This way you can just log everything to mlflow (instead of saving it locally) and dont worry about it  
Generally, inference is done in "production" where you don't have a labeled dataset (I wouldn't mix it with test)
We don't really need run_inference for now, but if Maxim insists that you implement it, I'd imagine it should use data_dir (for loading some dataset) and predictions_dir to save the predictions

File: src/models/BaseNextPatchForecaster.py [hunk 2]
```diff
@@ -16,6 +17,8 @@ def __init__(self, config: Dict[str, Any], patch_len: int):
         self.weight_decay = config["training"]["weight_decay"]
         self.patch_len = patch_len
         self.save_hyperparameters(config)
+        self.anomaly_threshold_train = 0.5
```
Comment (SharonPeled) [inline src/models/BaseNextPatchForecaster.py:20]: I'd start from None, and would log meaningless metrics for the first epoch

File: src/models/BaseNextPatchForecaster.py
Comment (SharonPeled) [inline src/models/BaseNextPatchForecaster.py:1]: process_batch_pred looks like code duplication. I think we should rely on the test process, as I mentioned earlier
Also, if I'm not wrong, the loss you save in _epoch_train_losses is the batch loss and you need the sample loss.

File: src/models/BaseNextPatchForecaster.py [hunk 3]
```diff
@@ -47,25 +50,39 @@ def configure_optimizers(self):
...
+        
+    def on_train_epoch_start(self):
+        self._epoch_train_losses = []
+
+    def on_train_epoch_end(self):
+        if len(self._epoch_train_losses) == 0:
+            return
+        per_step = torch.stack(self._epoch_train_losses)
+        all_losses = self.all_gather(per_step).reshape(-1).float().cpu().numpy()
+        self.anomaly_threshold_train = float(np.percentile(all_losses, 99))
```
Comment (SharonPeled) [inline src/models/BaseNextPatchForecaster.py]: The 99th percentile should be in config
